core/src/main/java/org/elasticsearch/index/mapper/TypeFieldMapper.java

```diff
@@ -123,49 +110,71 @@ public String typeName() {
             return CONTENT_TYPE;
         }
 
-        public boolean fielddata() {
-            return fielddata;
-        }
-
-        public void setFielddata(boolean fielddata) {
-            checkIfFrozen();
-            this.fielddata = fielddata;
-        }
-
         @Override
         public IndexFieldData.Builder fielddataBuilder() {
             if (hasDocValues()) {
                 return new DocValuesIndexFieldData.Builder();
+            } else {
+                // means the index has a single type and the type field is implicit
+                Function<MapperService, String> typeFunction = mapperService -> {
+                    Collection<String> types = mapperService.types();
+                    if (types.size() > 1) {
+                        throw new AssertionError();
+                    }
+                    String type = types.isEmpty() ? "doc" : types.iterator().next();
```

Comment:
Here we return "doc" if we don't have any types but in `termsQuery` we return `match_none`. Do we expect to get here when we don't have any types?